goldmane/pkg/aggregator/aggregator.go

```diff
@@ -0,0 +1,307 @@
+			// of the aggregated flows as the end time of this bucket.
+			logrus.WithField("index", i).Debug("No need to check remaining buckets")
+			break
+		}
+
+		// If this bucket's start time is after the end time of the request, then we can
+		// skip this bucket and move on to the next one.
+		if req.StartTimeLt > 0 && bucket.StartTime >= req.StartTimeLt {
+			logrus.WithField("index", i).Debug("Skipping bucket because it starts after the requested time window")
+			continue
+		}
+
+		// Check each flow in the bucket to see if it matches the request.
+		for key, flow := range bucket.Flows {
+			if !flowMatches(flow, req) {
+				logrus.Debug("Skipping flow because it doesn't match the request")
+				continue
+			}
+
+			if _, ok := flowsByKey[key]; !ok {
+				// Initialize the flow if it doesn't exist by making a copy.
+				cp := *flow
+				logrus.WithField("idx", i).WithFields(bucket.Fields()).Debug("Adding new flow to results")
+
+				// Set the start and end times of this flow to match the bucket.
+				// Aggregated flows always align with bucket intervals for consistent rate calculation.
+				cp.StartTime = bucket.StartTime
+				cp.EndTime = bucket.EndTime
+				flowsByKey[key] = &cp
+			} else {
+				logrus.WithField("idx", i).WithFields(bucket.Fields()).Debug("Adding flow contribution from bucket to results")
+
+				// Add this bucket's contribution to the flow.
+				mergeFlowInto(flowsByKey[key], flow)
+
+				// Since this flow was present in a later (chronologically) bucket, we need to update the start time
+				// of the flow to the start time of this (earlier chronologically) bucket.
+				flowsByKey[key].StartTime = bucket.StartTime
+			}
+		}
+	}
+
+	// Convert the map to a slice.
+	flows := []proto.Flow{}
+	for _, flow := range flowsByKey {
+		flows = append(flows, *types.FlowToProto(flow))
+	}
+	// Sort the flows by start time, sorting newer flows first.
+	sort.Slice(flows, func(i, j int) bool {
+		return flows[i].StartTime > flows[j].StartTime
+	})
+
+	// If pagination was requested, apply it now after sorting.
+	// This is a bit inneficient - we collect more data than we need to return -
+	// but it's a simple way to implement basic pagination.
+	if req.PageSize > 0 {
+		startIdx := (req.PageNumber) * req.PageSize
+		endIdx := startIdx + req.PageSize
+		if startIdx >= int64(len(flows)) {
+			return []proto.Flow{}
+		}
+		if endIdx > int64(len(flows)) {
+			endIdx = int64(len(flows))
+		}
+		flows = flows[startIdx:endIdx]
+		logrus.WithFields(logrus.Fields{
+			"pageSize":   req.PageSize,
+			"pageNumber": req.PageNumber,
+			"startIdx":   startIdx,
+			"endIdx":     endIdx,
+			"total":      len(flows),
+		}).Debug("Returning paginated flows")
+	}
+	return flows
+}
+
+func (a *LogAggregator) Stop() {
+	close(a.done)
+}
+
+func (a *LogAggregator) rollover() {
+	// Add a new bucket at the start and remove the last bucket.
+	start := time.Unix(a.buckets[0].EndTime, 0)
+	end := start.Add(a.rolloverTime)
```

Comment:
I have some small fear that small discrepancies in rollover time can cause our internal bucket array to eventually fall behind the current time.

If a rollover() is delayed by a second, for example due to a long query, then when the rollover occurs the new bucket will be offset by a second from the actual time. If this happens enough, then eventually new bucket end times may actually be in the past! We'll potentially get flow updates for `time.Now()` that don't match any of the buckets in our array. 

Need to think about how to resolve this - we could try to spot when this has happened and add buckets to compensate?

Reply:
I've got a local commit that addresses this by using a variable, calculated rollover time to make sure we catch up / slow down our rollovers as needed.